plugins/CoreHome/vue/src/MatomoCopyModal/MatomoCopyModalStore.ts

```diff
@@ -0,0 +1,148 @@
+/*!
+ * Matomo - free/libre analytics platform
+ *
+ * @link    https://matomo.org
+ * @license https://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+
+import { reactive } from 'vue';
+import { translateOrDefault } from '../translate';
+import Matomo from '../Matomo/Matomo';
+import MatomoUrl from '../MatomoUrl/MatomoUrl';
+
+interface MatomoCopyState {
+  /**
+   * Whether the modal is currently visible
+   */
+  isModalVisible: boolean
+  /**
+   * Whether the watching is suppressed. This allows updating the form data without validation.
+   */
+  isWatchSuppressed: boolean
+  /**
+   * Form data that needs to be included in the request sent to the server but won't change between
+   * requests. This could be a parent ID (e.g. ID of the container when copying MTM tags).
+   */
+  commonFormData?: Record<string, unknown>;
+  /**
+   * Additional form data that needs to be included in the request sent to the server. This should
+   * typically include the unique identifier of the entity being copied (e.g. idGoal for a goal).
+   */
+  entityFormData: Record<string, unknown>;
+  /**
+   * Should uniquely identify what is being copied (e.g. goal, funnel, segment, ...). The is
+   * important as it's used as the entityTypeName property of the request sent to the server.
+   */
+  copyEntityType: string;
+  /**
+   * Translation of what is being copied (e.g. goal, funnel, segment, ...). This can be a string
+   * or translation key. If nothing is provided 'report' is used.
+   */
+  copyEntityTypeTranslation: string;
+}
+
+export class MatomoCopyModalStore {
+  state: MatomoCopyState = reactive({
+    isModalVisible: false,
+    isWatchSuppressed: false,
+    commonFormData: {},
+    entityFormData: {},
+    copyEntityType: '',
+    copyEntityTypeTranslation: '',
+  });
+
+  constructor(
+    copyEntityType: string,
+    copyEntityTypeTranslation: string,
+    commonFormData?: Record<string, unknown>,
+  ) {
+    this.state.copyEntityType = copyEntityType;
+    this.state.copyEntityTypeTranslation = copyEntityTypeTranslation;
+    this.state.commonFormData = commonFormData ?? {};
+  }
+
+  showModal(entityFormData?: Record<string, unknown>): void {
+    // Make sure that we start fresh
+    this.resetFormData();
+
+    // Update the store with any provided form data
+    Object.entries(entityFormData ?? {}).forEach(([key, value]) => {
+      this.state.entityFormData[key] = value;
+    });
```

Comment:
I think we could use object destructuring here for the copy (to make it shorter)
```suggestion
    // Update the store with any provided form data
    this.state.entityFormData = { ... entityFormData };
```

Reply:
Thank you. That's a good suggestion. However, I purposely update the entityFormData rather than reassign it so that reactivity is maintained for all components using the store.